### DISCUSSION TITLE
Multiple-Key Setup with Easy Switching, even for same API provider

### DISCUSSION THREAD

**Discussion by JTimothy75** [cline/cline#869] [Feature Requests]
I would like to request the following enhancements for managing API keys:  

1. **Multi-Key Setup**
   - Enable me to add multiple API keys, both for office use and personal use.  
   - Allow assigning descriptive names to these keys for easier identification.  

2. **Seamless Key Switching**  
   - Provide a list of all saved setups (as illustrated in the attached image) and allow switching between them without the need to re-enter keys.  
   - Example list format:  
     - GPT-4o PNCL  
     - GPT-4o Mini PNCL  
     - Claude 3.5 Sonnet PNCL  
     - GPT-4o JTim  
     - GPT-4o Mini JTim  
     - Claude 3.5 Sonnet JTim  
     - Llama3.1 Chat  
     - Qwen-2.5 3b Chat  
     - Grok Beta JTim  

This functionality would streamline workflows by eliminating repetitive key entries and providing a centralized way to manage multiple API configurations.  

![Sample](https://github.com/user-attachments/assets/949771eb-9687-407b-b89e-abd93eb150c5)

**Comment by IsaiahJoseph** [comment]
That's great 👍

**Comment by Davidladan1050** [comment]
Great work 👍

**Reply by ghost** [reply]
Greatt work, keep going! 🔥

**Comment by yan5xu** [comment]
This is a must-have feature, as different models are suitable for different work scenarios.

**Comment by Gabbo7474** [comment]
That's great! And with the Requesty router extension that enables switching keys and models it would make very nice and easy to manage workflow!

**Comment by shlgug** [comment]
RooCode already supports this via multiple "profiles" that allow us to name each one and put in a different API key for each one. This is a CRITICAL feature when working on projects for multiple clients and we need to track API usage separately for billing purposes. Also, please make sure to keep the profile being used by each instance of VSCode separate so that we can run Cline for two different projects with two different API keys in two different VSCode windows at the same time without it mixing up keys. Also, it would be great to add the ability to lock down which profiles can be used by a specific project to prevent the developer and Cline from using the wrong keys for the wrong projects.

**Reply by paraddise** [reply]
Yes, for example we can use something like `.cline.env`  or `.clinerc` file for project where we can specify name of the existing profile to use or url, key, model for new profile

**Comment by paraddise** [comment]
I have my personal projects and want to use public models, but for my company projects I want to use corporate private models, So I need ability to switch model without entering base_url, api key again

**Comment by dollro** [comment]
great idea - would be used immediately!
Also consider this scenario: have openrouter API, but also a PRO API key of gemini. If gemini model is selected, i want to use the direct API key of the pro account of gemini, otherwise openrouter....

**Comment by rockyicer** [comment]
great! 

one openrouter API, switch to diff, model is quite important to my work.
even for the same script, I also want to switch the model from cline GUI instead of change the name every time from setup.
looking forward to this new funciton. excited!

**Comment by tobergicp** [comment]
I would also like to request this feature. 

It would also be great if the VSCode Extension would support (workspace) settings to configure cline, including the api key.

**Comment by MrVolcano** [comment]
The whole AI Model switching is a bit clunky. Multiple models under multiple providers. I'd like to see a system of "Starring" favorite models so that the ones we use regularly are easily switchable between with one click.

**Comment by Shayan-Emami** [comment]
I Really need this.
it would be great if we can have that. 

currently i need it for personal and office use.

**Comment by es3154** [comment]
I Really need this.

**Comment by dbstraight** [comment]
This would be very very useful. I switch models all the time when using copilot.

**Comment by realalexandergeorgiev** [comment]
its not going to happen, is it?

**Reply by sTiKyt** [reply]
but why?

**Comment by DesterCalibra** [comment]
I can't express how useful this would be. I would often want to switch between my locally running LLM and, for example, Moonshot/Kimi. With one (or two maximum) clicks. Without the need to edit provider, key, etc.
